blob/service.go

```diff
@@ -0,0 +1,278 @@
+package blob
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/celestiaorg/celestia-app/pkg/shares"
+	apptypes "github.com/celestiaorg/celestia-app/x/blob/types"
+	libhead "github.com/celestiaorg/go-header"
+	"github.com/celestiaorg/nmt/namespace"
+
+	"github.com/celestiaorg/celestia-node/header"
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/state"
+)
+
+var (
+	log = logging.Logger("blob")
+
+	errBlobNotFound = errors.New("blob: blob not found")
+)
+
+type Service struct {
+	// accessor dials the given celestia-core endpoint to submit blobs.
+	accessor *state.CoreAccessor
+	// sGetter retrieves the EDS to fetch all shares from the requested header.
+	sGetter share.Getter
+	// hGetter allows to get the requested header.
+	hGetter libhead.Getter[*header.ExtendedHeader]
+	// headGetter gets current network head.
+	headGetter libhead.Head[*header.ExtendedHeader]
+}
+
+func NewService(
+	state *state.CoreAccessor,
+	getter share.Getter,
+	hGetter libhead.Getter[*header.ExtendedHeader],
+	headGetter libhead.Head[*header.ExtendedHeader],
+) *Service {
+	return &Service{
+		accessor:   state,
+		sGetter:    getter,
+		hGetter:    hGetter,
+		headGetter: headGetter,
+	}
+}
+
+func (s *Service) Submit(ctx context.Context, f int64, gasLimit uint64, blobs ...*Blob) (uint64, error) {
+	log.Debugw("submitting blobs.", "amount", len(blobs))
+
+	var (
+		fee = types.NewInt(f)
+		b   = make([]*apptypes.Blob, len(blobs))
+		err error
+	)
+
+	for i, blob := range blobs {
+		b[i], err = apptypes.NewBlob(blob.NamespaceId, blob.Data())
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	resp, err := s.accessor.SubmitPayForBlob(ctx, fee, gasLimit, b...)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(resp.Height), nil
```

Comment:
Agree with Hlib here. Would avoid returning TxResponse. It will require users to write their own parser for the response. We can implement it on our side